### ISSUE TITLE
[apex] ApexUnitTestClassShouldHaveAsserts additionalAssertMethodPattern not matching

### ISSUE THREAD

**Issue by kknapp1** [pmd/pmd#4680]
**Affects PMD Version:** 
6.55.0

**Rule:**
[ApexUnitTestClassShouldHaveAsserts](https://docs.pmd-code.org/latest/pmd_rules_apex_bestpractices.html#apexunittestclassshouldhaveasserts)

**Description:**
additionalAssertMethodPattern is not matching "verify" methods specified by the pattern. I'm sure it's user error here but could someone help me find what I'm doing wrong? 

I started by using the regex suggested here (https://github.com/pmd/pmd/issues/1089). I've tried variations on the regex that match in other regex testers and PMD reports all of them as failures.

**Code Sample demonstrating the issue:**
Apex Test
```java
@IsTest
public class Test {
   @IsTest
   static void doSomeWork_VerifyServiceIscalled() {
      final Integer testWorkValue = -1;
      fflib_ApexMocks mocks = new fflib_ApexMocks();
      FinancialService mockFinancialService = (FinancialService)mocks.mock(FinancialService.class);
      mocks.startStubbing();
      mocks.when(mockFinancialService.doWork()).thenReturn(testWorkValue);
      mocks.stopStubbing();

      Test.startTest();
      PersonService sut = new PersonService();
      sut.doSomeWork(mockFinancialService);
      Test.stopTest();

      ((FinancialService)mocks.verify(mockFinancialService, mocks.times(1))).doWork();
   }
}
```
Rule in ruleset
```xml
  <rule ref="category/apex/bestpractices.xml/ApexUnitTestClassShouldHaveAsserts" message="Test should have verify (my custom error message)">
    <priority>1</priority>
    <properties>
      <property name="additionalAssertMethodPattern" value="(Assert\.\w+|verify\w+)" />
    </properties>
  </rule>
```

I've confirmed that PMD correctly ignores other cases of "verify" method. But this usage of the fflib_apexMocks verify instance is used throughout our codebase and I could use help with a solution.

**Expected outcome:**
No PMD violations.

**Actual outcome:**
PMD reports a violation, but that's wrong. That's a false positive.

**Running PMD through:** *[CLI]*

**Comment by adangel** [comment]
Maybe we match the whole method name? Then your regex needs to be

```
Assert\.\w+|verify
```

note: without the "\w+" at the end, as the method is just called verify without additional characters

**Comment by kknapp1** [comment]
Thank you for the suggestion @adangel but that regex doesn't seem to fix the problem. PMD still doesn't recognize the "verify" method call and throws an error.

I have tried several variations on the regex with no success. Testing the regex patterns at https://regex101.com/ are successful, but still not matching by PMD. I realize regex parsers can vary - is there a recommended parser that will effectively emulate PMDs regex matcher?

**Comment by adangel** [comment]
So, the rule is actually checking the "full method name": https://github.com/pmd/pmd/blob/347d5992d13be82418cc1c7018bf6f7e4db7853d/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestClassShouldHaveAssertsRule.java#L112

In your example, the full method name is `mocks.verify`, that means, your regex needs to be `mocks\.verify`. In case your variable name "mocks" is different in each test case, you might also use `.*\.verify` instead.

Also note, that `Assert\.\w+` is not needed, because the rule already checks for that: https://github.com/pmd/pmd/blob/347d5992d13be82418cc1c7018bf6f7e4db7853d/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestClassShouldHaveAssertsRule.java#L40-L48

The regex is for _additional_ assert-like methods.

In summary, your rule configuration should look like this:

```xml
  <rule ref="category/apex/bestpractices.xml/ApexUnitTestClassShouldHaveAsserts" message="Test should have verify (my custom error message)">
    <priority>1</priority>
    <properties>
      <property name="additionalAssertMethodPattern" value=".+\.verify" />
    </properties>
  </rule>
```

**Comment by kknapp1** [comment]
Yes! Thank you very much for the solution @adangel ! And also for the links to the java files. That helps me get oriented to the codebase and I can hopefully answer my own questions in the future. Appreciate all you do!
